pkg/virt-operator/install-strategy/create.go

```diff
@@ -0,0 +1,996 @@
+		return false
+	}
+
+	foundImageTag := objectMeta.Annotations[v1.InstallStrategyVersionAnnotation]
+	foundImageRegistry := objectMeta.Annotations[v1.InstallStrategyRegistryAnnotation]
+
+	if foundImageTag == imageTag && foundImageRegistry == imageRegistry {
+		return true
+	}
+
+	return false
+}
+
+func apiDeployments(strategy *InstallStrategy) []*appsv1.Deployment {
+	var deployments []*appsv1.Deployment
+
+	for _, deployment := range strategy.deployments {
+		if !strings.Contains(deployment.Name, "virt-api") {
+			continue
+		}
+		deployments = append(deployments, deployment)
+
+	}
+	return deployments
+}
+
+func controllerDeployments(strategy *InstallStrategy) []*appsv1.Deployment {
+	var deployments []*appsv1.Deployment
+
+	for _, deployment := range strategy.deployments {
+		if strings.Contains(deployment.Name, "virt-api") {
+			continue
+		}
+		deployments = append(deployments, deployment)
+
+	}
+	return deployments
+}
+
+func injectOperatorLabelAndAnnotations(objectMeta *metav1.ObjectMeta, imageTag string, imageRegistry string) {
+	if objectMeta.Labels == nil {
+		objectMeta.Labels = make(map[string]string)
+	}
+	objectMeta.Labels[v1.ManagedByLabel] = v1.ManagedByLabelOperatorValue
+
+	if objectMeta.Annotations == nil {
+		objectMeta.Annotations = make(map[string]string)
+	}
+	objectMeta.Annotations[v1.InstallStrategyVersionAnnotation] = imageTag
+	objectMeta.Annotations[v1.InstallStrategyRegistryAnnotation] = imageRegistry
+}
+
+func generatePatchBytes(ops []string) []byte {
+	opsStr := "["
+	for idx, entry := range ops {
+		sep := ", "
+		if len(ops)-1 == idx {
+			sep = "]"
+		}
+		opsStr = fmt.Sprintf("%s%s%s", opsStr, entry, sep)
+	}
+	return []byte(opsStr)
+}
+
+func createLabelsAndAnnotationsPatch(objectMeta *metav1.ObjectMeta) ([]string, error) {
+	var ops []string
+	labelBytes, err := json.Marshal(objectMeta.Labels)
+	if err != nil {
+		return ops, err
+	}
+	annotationBytes, err := json.Marshal(objectMeta.Annotations)
+	if err != nil {
+		return ops, err
+	}
+	ops = append(ops, fmt.Sprintf(`{ "op": "add", "path": "/metadata/labels", "value": %s }`, string(labelBytes)))
+	ops = append(ops, fmt.Sprintf(`{ "op": "add", "path": "/metadata/annotations", "value": %s }`, string(annotationBytes)))
+
+	return ops, nil
+}
+
+func syncDaemonSet(kv *v1.KubeVirt,
+	daemonSet *appsv1.DaemonSet,
+	stores util.Stores,
+	clientset kubecli.KubevirtClient,
+	expectations *util.Expectations) error {
+
+	daemonSet = daemonSet.DeepCopy()
+
+	apps := clientset.AppsV1()
+	imageTag := kv.Status.TargetKubeVirtVersion
+	imageRegistry := kv.Status.TargetKubeVirtRegistry
+
+	injectOperatorLabelAndAnnotations(&daemonSet.ObjectMeta, imageTag, imageRegistry)
+	injectOperatorLabelAndAnnotations(&daemonSet.Spec.Template.ObjectMeta, imageTag, imageRegistry)
+
+	kvkey, err := controller.KeyFunc(kv)
+	if err != nil {
+		return err
+	}
+
+	var cachedDaemonSet *appsv1.DaemonSet
+	obj, exists, _ := stores.DaemonSetCache.Get(daemonSet)
+	if exists {
```

Comment:
yeah, I picked the pattern and just went with it everywhere, including the SyncAll function that we can't 'return' early out of. Since this function is a one off to help out in update vs rollback path branching logic, I could revise the logic.